### PR TITLE
[JENKINS-755] Describe the built-in JDK as '(System)'.

### PR DESCRIPTION
**Pull request by Wilfred**
As suggested in
https://issues.jenkins-ci.org/browse/JENKINS-755?focusedCommentId=206175&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-206175
this avoids confusion where users incorrectly think they're getting a
JDK they've listed in the configuration page.

### CODE REVIEW

**Comment by daniel-beck** [comment]
Needs to add backwards compatibility code to not issue tons of bogus warnings like https://github.com/jenkinsci/jenkins/commit/a2341e6b4c854a0cc183b601acdba7572ad79152 (for all projects not having been reconfigured since it was named `(Default)`.

**Comment by daniel-beck** [comment]
Related PR: #1451.

**Comment by Wilfred** [comment]
@daniel-beck OK, I've pushed a fix to avoid generating warnings. I'm not familiar with the Jenkins codebase, so let me know if you'd prefer a different approach.

File: core/src/main/java/hudson/model/JDK.java [hunk 1]
```diff
@@ -54,10 +54,22 @@
...
 
     /**
-     * Name of the “default JDK”, meaning no specific JDK selected.
+     * Name of the “System JDK”, which is just the JDK on Jenkins' $PATH.
      * @since 1.577
      */
-    public static final String DEFAULT_NAME = "(Default)";
+    public static final String DEFAULT_NAME = "(System)";
+
+    public static boolean isDefaultName(String name) {
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/JDK.java]: Consider `@Restricted(NoExternalUse.class)`, I see no reason to make this public API.

Comment (Wilfred) [inline core/src/main/java/hudson/model/JDK.java]: Fixed.

**Comment by Wilfred** [comment]
@daniel-beck this should now be good to go -- are you happy?

File: core/src/main/java/hudson/model/JDK.java [hunk 2]
```diff
@@ -54,10 +56,23 @@
...
-     * Name of the “default JDK”, meaning no specific JDK selected.
+     * Name of the “System JDK”, which is just the JDK on Jenkins' $PATH.
      * @since 1.577
      */
-    public static final String DEFAULT_NAME = "(Default)";
+    public static final String DEFAULT_NAME = "(System)";
+
+    @Restricted(NoExternalUse.class)
+    public static boolean isDefaultName(String name) {
+        if (name != null) {
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/JDK.java]: The else block will never be entered. Is this supposed to be `==`?

Comment (Wilfred) [inline core/src/main/java/hudson/model/JDK.java]: Woops! Fixed.

Comment (daniel-beck) [inline core/src/main/java/hudson/model/JDK.java]: Should be trivial to write a test for this method.

Comment (Wilfred) [inline core/src/main/java/hudson/model/JDK.java]: Added test and refactored the logic a little.

**Comment by Wilfred** [comment]
@daniel-beck tells me he's reluctant to merge this because #1451 is more comprehensive. Note that #1436 is almost exactly the same at this.

**Comment by oleg-nenashev** [comment]
Now it competes with https://github.com/jenkinsci/jenkins/pull/1451
I prefer this one, because it has tests

**Comment by oleg-nenashev** [comment]
#1451  has been closed by @ndeloof . I tend to :+1: regarding the merge

**Comment by oleg-nenashev** [comment]
Marked as ready-for-merge

**Comment by oleg-nenashev** [comment]
@Wilfred 
Thanks a lot for your contribution!

File: core/src/test/java/jenkins/model/JDKNameTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2014
```
Comment (recena) [inline core/src/test/java/jenkins/model/JDKNameTest.java]: Maybe `2015`?

Comment (oleg-nenashev) [inline core/src/test/java/jenkins/model/JDKNameTest.java]: Probably. BTW it's up to the commit author. Probably he developed this code in 2014

File: core/src/main/java/hudson/model/JDK.java [hunk 4]
```diff
@@ -54,10 +56,19 @@
...
-    public static final String DEFAULT_NAME = "(Default)";
+    public static final String DEFAULT_NAME = "(System)";
+
+    @Restricted(NoExternalUse.class)
+    public static boolean isDefaultName(String name) {
+        if ("(Default)".equals(name)) {
+            // DEFAULT_NAME took this value prior to 1.598.
+            return true;
+        }
+        return DEFAULT_NAME.equals(name);
```
Comment (jglick) [inline core/src/main/java/hudson/model/JDK.java]: [JENKINS-31217](https://issues.jenkins-ci.org/browse/JENKINS-31217):

```java
|| name == null
```

File: core/src/test/java/jenkins/model/JDKNameTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.jvnet.hudson.test.Issue;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+@RunWith(PowerMockRunner.class)
+public class JDKNameTest {
+    @Test
+    public void nullIsNotDefaultName() {
+        assertThat(JDK.isDefaultName(null), is(false));
```
Comment (jglick) [inline core/src/test/java/jenkins/model/JDKNameTest.java]: JENKINS-31217: wrong.

**Comment by jglick** [comment]
Would have been better to just modify the display name in the tool dropdown, I think, rather than actually force a settings change. (The use of the display name in settings to begin with was clearly a mistake; should have used some code token like `_SYSTEM_`.)

**Comment by daniel-beck** [comment]
@Wilfred Please let us know whether you plan to amend this fix in a new PR, ideally this week.

**Pull request merged**